packages/serve/test/middlewares/built-in-middlewares/authRouteMiddleware.spec.ts

```diff
@@ -0,0 +1,211 @@
+      []
+    );
+    // spy the async function to do test
+    const spy = jest.spyOn(lodash, 'isEmpty');
+
+    await middleware.handle(ctx, async () => Promise.resolve());
+
+    expect(spy).not.toHaveBeenCalled();
+  });
+
+  it.each([[{}], [undefined]])(
+    'Should throw error when options = %p',
+    async (options) => {
+      // Arrange
+      const expected = new Error(
+        'please set at least one auth type and user credential when you enable the "auth" options.'
+      );
+      const ctx: KoaContext = {
+        ...sinon.stubInterface<KoaContext>(),
+      };
+      // Act
+      const middleware = new AuthRouteMiddleware({ options: options }, '', []);
+
+      const action = middleware.handle(ctx, async () => Promise.resolve());
+
+      expect(action).rejects.toThrow(expected);
+    }
+  );
+
+  it('Should active corresponding authenticator when activate middleware', async () => {
+    // Arrange
+    const middleware = new AuthRouteMiddleware(
+      {},
+      '',
+      Object.keys(stubAuthenticators).map((name) => stubAuthenticators[name])
+    );
+
+    // Act
+    await middleware.activate();
+
+    // Assert
+    expect(stubAuthenticators['a1'].activate?.called).toBeTruthy();
+    expect(stubAuthenticators['a2'].activate?.called).toBeTruthy();
+  });
+
+  it('Should failed when activate with non-supported authenticator id "m1"', async () => {
+    // Arrange
+    const expected = new Error(
+      `The auth type "m1" in options not supported, authenticator only supported ${Object.keys(
+        stubAuthenticators
+      )}.`
+    );
+
+    const middleware = new AuthRouteMiddleware(
+      { options: { a1: {}, m1: {} } },
+      '',
+      Object.keys(stubAuthenticators).map((name) => stubAuthenticators[name])
+    );
+
+    // Act
+    const activate = middleware.activate();
+    expect(activate).rejects.toThrowError(expected);
+  });
+
+  it.each([
+    ['a1', 'a1-token'],
+    ['a2', 'a2-token'],
+  ])(
+    'Should get correct result when call GET /auth?type=%p',
+    async (name, expected) => {
+      // Arrange
+      const server = await runServer({ a1: {}, a2: {} }, stubAuthenticators);
+
+      // Act
+      const request = supertest(server).get(`/auth?type=${name}`);
+      const response = await request;
+      // Assert
+      expect(response.body).toEqual({ token: expected });
+      server.close();
+    }
+  );
+
+  it('Should failed when call GET /auth but not include "type" query string', async () => {
+    // Arrange
+    const options = { a1: {}, a2: {} };
+    const expected = {
+      message: `Please indicate auth "type", supported auth types: ${Object.keys(
+        options
+      )}.`,
+    };
+    const server = await runServer(options, stubAuthenticators);
+
+    // Act
+    const request = supertest(server).get(`/auth`);
+    const response = await request;
+
+    // Assert
+    expect(response.statusCode).toEqual(400);
+    expect(response.body).toEqual(expected);
+    server.close();
```

Comment:
We should close server in afterEach hooks, or they become dangling if test cases failed.

Reply:
Thanks for @oscar60310 suggestion. it has been fixed.